db/compaction/compaction_job.cc

```diff
@@ -290,6 +296,128 @@ void CompactionJob::Prepare() {
+      // max_subcompaction. If yes, we need to compute the extra subcompaction
+      // based on compaction limits (the extra subcompactions should not
+      // be larger than that)
+      int extra_num_subcompactions_limit =
+          std::min((int)(num_planned_subcompactions - max_subcompactions_limit),
+                   available_bg_compactions_db_limit_);
+
+      TEST_SYNC_POINT_CALLBACK(
+          "CompactionJob::GetNumPlannedSubcompactions:"
+          "ExtraCompactionsAgainstDBLimits",
+          &extra_num_subcompactions_limit);
+
+      // Step 3: Try to reserve threads for extra subcompactions. The planned
+      // subcompactions may change again if we cannot obtain sufficient threads
+      extra_num_subcompaction_threads_reserved_ =
+          env_->ReserveThreads(extra_num_subcompactions_limit, thread_pri_);
+      num_planned_subcompactions =
+          static_cast<uint64_t>(extra_num_subcompaction_threads_reserved_) +
+          max_subcompactions_limit;
+      TEST_SYNC_POINT_CALLBACK(
+          "CompactionJob::GetNumPlannedSubcompactions:Finalize",
+          &num_planned_subcompactions);
+    }
+  }
+  return num_planned_subcompactions;
+}
+
+void CompactionJob::UpdateResourcesAndLimits(
+    uint64_t num_actual_subcompactions, uint64_t num_planned_subcompactions) {
+  // We only need to update resources and limits if the actual subcompactions
+  // is less than the planned one and we did reserve some resources earlier
+  if (extra_num_subcompaction_threads_reserved_ > 0 &&
+      num_actual_subcompactions < num_planned_subcompactions) {
+    // We cannot release threads more than what we reserved before
+    int additional_bg_compactions = env_->ReleaseThreads(
+        std::min((int)(num_planned_subcompactions - num_actual_subcompactions),
+                 extra_num_subcompaction_threads_reserved_),
+        thread_pri_);
+    // Update the number of reserved threads and the number of background
+    // scheduled compactions for this compaction job
+    extra_num_subcompaction_threads_reserved_ -= additional_bg_compactions;
+    db_mutex_->AssertHeld();
+    // Update bg_compaction_scheduled_ or bg_bottom_compaction_scheduled_
+    // depending on if this compaction has the bottommost priority
+    if (thread_pri_ == Env::Priority::BOTTOM && bottommost_level_) {
+      if (bg_bottom_compaction_scheduled_ != nullptr) {
+        *bg_bottom_compaction_scheduled_ +=
+            extra_num_subcompaction_threads_reserved_;
+      }
+    } else if (bg_compaction_scheduled_ != nullptr) {
+      *bg_compaction_scheduled_ += extra_num_subcompaction_threads_reserved_;
+    }
+    TEST_SYNC_POINT("CompactionJob::UpdateResourcesAndLimits:0");
+  }
+}
+
+void CompactionJob::ReleaseResourcesAndLimits() {
+  // The number of reserved threads become larger than 0 only if the
+  // compaction prioity is round robin and there is no sufficient
+  // sub-compactions available
+  if (extra_num_subcompaction_threads_reserved_ > 0) {
```

Comment:
Will `extra_num_subcompaction_threads_reserved_` ever be negative? 
If not, would it be fine to leave `std::min` to take care of `extra_num_subcompaction_threads_reserved_ == 0` case and remove one layer of nested if-else?

Reply:
I can add an early return when `extra_num_subcompaction_threads_reserved_ == 0` here so that we avoid the lock.

Reply:
Do you mean avoiding the lock in the early return when `extra_num_subcompaction_threads_reserved_ == 0`?

Reply:
Yes.